src/app/chat/views/user_list.nim

```diff
@@ -0,0 +1,118 @@
+import NimQml, Tables, json, chronicles, sequtils
+import ../../../status/status
+import ../../../status/accounts
+import ../../../status/chat
+import ../../../status/chat/[message]
+import strutils
+
+type
+  UserListRoles {.pure.} = enum
+    UserName = UserRole + 1
+    LastSeen = UserRole + 2
+    PublicKey = UserRole + 3
+    Alias = UserRole + 4
+    LocalName = UserRole + 5
+    Identicon = UserRole + 6
+
+  User = object
+    username: string
+    alias: string
+    localName: string
+    lastSeen: string
+    identicon: string
+
+QtObject:
+  type
+    UserListView* = ref object of QAbstractListModel
+      status: Status
+      users: seq[string]
+      userDetails: OrderedTable[string, User]
+
+  proc delete(self: UserListView) =
+    self.userDetails.clear()
+    self.QAbstractListModel.delete
+
+  proc setup(self: UserListView) =
+    self.QAbstractListModel.setup
+
+  proc newUserListView*(status: Status): UserListView =
+    new(result, delete)
+    result.userDetails = initOrderedTable[string, User]()
+    result.users = @[]
+    result.status = status
+    result.setup
+
+  method rowCount*(self: UserListView, index: QModelIndex = nil): int = self.users.len
+
+  method data(self: UserListView, index: QModelIndex, role: int): QVariant =
+    if not index.isValid:
+      return
+    if index.row < 0 or index.row >= self.users.len:
+      return
+    
+    let user = self.users[index.row]
+
+    case role.UserListRoles:
+      of UserListRoles.UserName: result = newQVariant(self.userDetails[user].userName)
+      of UserListRoles.LastSeen: result = newQVariant(self.userDetails[user].lastSeen)
+      of UserListRoles.Alias: result = newQVariant(self.userDetails[user].alias)
+      of UserListRoles.LocalName: result = newQVariant(self.userDetails[user].localName)
+      of UserListRoles.PublicKey: result = newQVariant(user)
+      of UserListRoles.Identicon: result = newQVariant(self.userdetails[user].identicon)
+
+  method roleNames(self: UserListView): Table[int, string] =
+    {
+      UserListRoles.UserName.int:"userName",
+      UserListRoles.LastSeen.int:"lastSeen",
+      UserListRoles.PublicKey.int:"publicKey",
+      UserListRoles.Alias.int:"alias",
+      UserListRoles.LocalName.int:"localName",
+      UserListRoles.Identicon.int:"identicon"
+    }.toTable
+
+  proc add*(self: UserListView, message: Message) =
+    if self.userDetails.hasKey(message.fromAuthor):
+        self.beginResetModel()
```

Comment:
I don't think this is the right thing to call. Reset is to be called when we change the whole model. You need to use `dataChanged` I think

Reply:
I'm using resetModel because DelegateModelGeneralized only sorts records that are added/removed.
In this case I am updating a record, and want to have its position in the list of users change depending on the lastSeen value.

Perhaps there's a way to call the .sort() function of the delegateModelGeneralized when dataChanged signal is triggered?